src/S3/MultipartCopy.php

```diff
@@ -127,16 +130,20 @@ private function createPart($partNumber, $partsCount)
         foreach ($params as $k => $v) {
             $data[$k] = $v;
         }
-
-        list($bucket, $key) = explode('/', ltrim($this->source, '/'), 2);
-        $data['CopySource'] = '/' . $bucket . '/' . implode(
-            '/',
-            array_map(
-                'urlencode',
-                explode('/', rawurldecode($key))
-            )
-        );
+        // The source parameter here is usually a string, but can be overloaded as an array
+        // if the key contains a '?' character to specify where the query parameters start
+        if (is_array($this->source)) {
+            $key = str_replace('%2F', '/', rawurlencode($this->source['source_key']));
+            $data['CopySource'] = '/' . $this->source['source_bucket'] . '/' . $key;
+        } else {
+            list($bucket, $key) = explode('/', ltrim($this->source, '/'), 2);
```

Comment:
Does this `else` block provide the exact same values as the previous code in all situations? I don't think the integration tests are validating that this is actually the same result, more that it doesn't break the process overall.